modules/wallet/database.go

```diff
@@ -38,91 +68,115 @@ func dbDelete(b *bolt.Bucket, key interface{}) error {
 	return b.Delete(encoding.Marshal(key))
 }
 
-// dbPutHistoricClaimStart stores the number of siafunds corresponding to the
-// specified SiafundOutputID. The wallet builds a mapping of output -> value
-// so that it can determine the value of siafund outputs that were not created
-// by the wallet.
+// dbForEach is a helper function for iterating over a bucket and calling fn
+// on each entry. fn must be a function with two parameters. The key/value
+// bytes of each bucket entry will be unmarshalled into the types of fn's
+// parameters.
+func dbForEach(b *bolt.Bucket, fn interface{}) error {
+	// check function type
+	fnVal, fnTyp := reflect.ValueOf(fn), reflect.TypeOf(fn)
+	if fnTyp.Kind() != reflect.Func || fnTyp.NumIn() != 2 {
+		panic("bad fn type: needed func(key, val), got " + fnTyp.String())
+	}
+
+	return b.ForEach(func(keyBytes, valBytes []byte) error {
+		key, val := reflect.New(fnTyp.In(0)), reflect.New(fnTyp.In(1))
+		if err := encoding.Unmarshal(keyBytes, key.Interface()); err != nil {
+			return err
+		} else if err := encoding.Unmarshal(valBytes, val.Interface()); err != nil {
+			return err
+		}
+		fnVal.Call([]reflect.Value{key.Elem(), val.Elem()})
+		return nil
+	})
+}
```

Comment:
I don't like this function, I think I would prefer to see a `ForEach` defined on each type that gets used in the codebase, similar to the other helper functions in this file. The reflection here is complicated, and is performed in a loop.

Maybe it's just because I'm pretty unfamiliar with reflection, but I don't find this code to be readable.